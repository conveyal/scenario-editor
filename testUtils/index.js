export function mockComponents (components) {
  if (components instanceof Array) {
    const mockedComponents = {}
    for (let i = 0; i < components.length; i++) {
      const name = components[i]
      mockedComponents[name] = name
    }
    return mockedComponents
  }
}
