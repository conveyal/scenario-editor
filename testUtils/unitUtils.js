/* globals describe, expect, it */

import React from 'react'
import renderer from 'react-test-renderer'

/**
 * Performs a basic render test of a Component
 *
 * @param  {Mixed} components Can be either an array of objects or a single object
 *   Each object can have the following properties:
 *   - children: Any children to render within the component
 *   - component (required): The component to render
 *   - name: How to classify the test in the `describe section`
 *   - notToBeCalledFns: An array of String keys that map to property values that are jest.fn()'s that will be asserted as not called during basic render test
 *   - props (required): The props of the component to render
 *   - toBeCalledFns: Same as `notToBeCalledFns` except they are asserted to be called
 */
export function basicRenderTest (components) {
  // prepare components to be tested
  if (!Array.isArray(components)) {
    components = [components]
  }

  // run tests on each component
  components.forEach((component) => {
    const Component = component.component

    if (typeof component.name !== 'string') {
      throw new Error('Component name not provided in basicRenderTest cfg')
    }

    describe(`Component > ${component.name}`, () => {
      it('renders correctly', () => {
        const tree = renderer.create(
          <Component
            {...component.props}
            >
            {component.children}
          </Component>
        ).toJSON()
        expect(tree).toMatchSnapshot()

        if (Array.isArray(component.toBeCalledFns)) {
          component.toBeCalledFns.forEach((fn) => {
            expect(component.props[fn]).toBeCalled()
          })
        }

        if (Array.isArray(component.notToBeCalledFns)) {
          component.notToBeCalledFns.forEach((fn) => {
            expect(component.props[fn]).not.toBeCalled()
          })
        }
      })
    })
  })
}

/**
 * Creates a object to mock a module with multiple exports
 *
 * @param  {Array} components An array of Strings representing the named exports
 * @return {Object}           A mock representation of the named exports
 */
export function mockExports (components) {
  if (Array.isArray(components)) {
    const mockedComponents = {}
    for (let i = 0; i < components.length; i++) {
      const name = components[i]
      mockedComponents[name] = name
    }
    return mockedComponents
  }
}
