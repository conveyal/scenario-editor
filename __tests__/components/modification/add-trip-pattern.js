/* global describe, it, expect, jest */

import React from 'react'
import renderer from 'react-test-renderer'

import { mockComponents } from '../../../testUtils'

jest.mock('../../../lib/components/buttons', () => { return mockComponents(['Button']) })
jest.mock('../../../lib/components/icon', () => 'Icon')
jest.mock('../../../lib/components/input', () => { return mockComponents(['Checkbox', 'Number']) })
jest.mock('../../../lib/components/timetable', () => 'Timetable')

import AddTripPattern from '../../../lib/components/modification/add-trip-pattern'

describe('AddTripPattern', () => {
  it('renders correctly', () => {
    const mapState = {
      allowExtend: true,
      extendFromEnd: true,
      followRoad: true,
      state: 'add-trip-pattern',
      modificationId: '1234'
    }
    const modification = {
      id: '1234',
      segments: [],
      timetables: [
        {
          name: 'Test timetable',
          speed: 40,
          dwellTime: 10,
          monday: true,
          tuesday: true,
          wednesday: true,
          thursday: true,
          friday: true,
          saturday: false,
          sunday: false,
          headwaySecs: 900, // 15 minutes
          startTime: 28800, // 8am
          endTime: 57600 // 4pm
        }
      ],
      bidirectional: false
    }
    const replaceModificationFn = jest.fn()
    const setMapStateFn = jest.fn()
    const updateFn = jest.fn()
    const tree = renderer.create(
      <AddTripPattern
        mapState={mapState}
        modification={modification}
        replaceModification={replaceModificationFn}
        setMapState={setMapStateFn}
        update={updateFn}
        />
    ).toJSON()
    expect(tree).toMatchSnapshot()
    expect(replaceModificationFn).not.toBeCalled
    expect(setMapStateFn).not.toBeCalled
    expect(updateFn).not.toBeCalled
  })
})
