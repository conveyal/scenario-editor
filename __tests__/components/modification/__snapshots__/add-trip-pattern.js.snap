exports[`Component > Modification > AddTripPattern renders correctly 1`] = `
<div>
  <Button
    block={true}
    onClick={[Function anonymous]}
    style="warning">
    <Icon
      type="times-circle" />
     Stop editing
  </Button>
  <Checkbox
    defaultChecked={true}
    label="Extend from end"
    onChange={[Function anonymous]} />
  <Checkbox
    defaultChecked={true}
    label="Follow streets"
    onChange={[Function anonymous]} />
  <Checkbox
    defaultChecked={true}
    label="Create stops automatically"
    onChange={[Function anonymous]} />
  <Number
    label="Stop spacing (meters)"
    onChange={[Function anonymous]}
    value={400} />
  <Checkbox
    checked={false}
    label="Bidirectional"
    onChange={[Function anonymous]} />
  <Timetable
    index={1}
    modification={
      Object {
        "bidirectional": false,
        "id": "1234",
        "segments": Array [],
        "timetables": Array [
          Object {
            "dwellTime": 10,
            "endTime": 57600,
            "friday": true,
            "headwaySecs": 900,
            "monday": true,
            "name": "Test timetable",
            "saturday": false,
            "speed": 40,
            "startTime": 28800,
            "sunday": false,
            "thursday": true,
            "tuesday": true,
            "wednesday": true
          }
        ]
      }
    }
    removeTimetable={[Function bound ]}
    replaceTimetable={[Function bound ]}
    timetable={
      Object {
        "dwellTime": 10,
        "endTime": 57600,
        "friday": true,
        "headwaySecs": 900,
        "monday": true,
        "name": "Test timetable",
        "saturday": false,
        "speed": 40,
        "startTime": 28800,
        "sunday": false,
        "thursday": true,
        "tuesday": true,
        "wednesday": true
      }
    } />
  <Button
    block={true}
    onClick={[Function anonymous]}
    style="success">
    <Icon
      type="plus" />
     Add timetable
  </Button>
</div>
`;
